torch/_tensor_docs.py

```diff
@@ -3120,14 +3120,20 @@ def callable(a, b) -> number
 
 This is the reverse operation of the manner described in :meth:`~Tensor.gather`.
 
-:attr:`self`, :attr:`index` and :attr:`src` (if it is a Tensor) should have same
-number of dimensions. It is also required that ``index.size(d) <= src.size(d)``
-for all dimensions ``d``, and that ``index.size(d) <= self.size(d)`` for all
-dimensions ``d != dim``.
+:attr:`self`, :attr:`index` and :attr:`src` (if it is a Tensor) should all have
+the same number of dimensions. It is also required that
+``index.size(d) <= src.size(d)`` for all dimensions ``d``, and that
+``index.size(d) <= self.size(d)`` for all dimensions ``d != dim``.
+Note that ``index`` and ``src`` do not broadcast.
 
 Moreover, as for :meth:`~Tensor.gather`, the values of :attr:`index` must be
-between ``0`` and ``self.size(dim) - 1`` inclusive, and all values in a row
-along the specified dimension :attr:`dim` must be unique.
+between ``0`` and ``self.size(dim) - 1`` inclusive.
+
+.. warning::
+
+    When indices are not unique, the behaviour is non-deterministic (one of the
+    values from ``src`` will be picked arbitrarily) and the gradient will be
+    incorrect!
```

Comment:
questions to @nikitaved as he implemented this
What is the gradient behavior here? Is it propagated to all source values instead of just one?
Are the cases where index size is smaller than self/src correctly handled now in the backward?

Reply:
I did not modify the backward behavior, so it is the same. Since it is `gather` in forward and it has a restriction that `index.shape == src.shape`, the backward is well-defined shape-wise. However, since it is `gather`, I believe its behavior is well-defined for non-unique indices. This warning is meant for `scatter` I presume. CC @rgommers

Reply:
Thanks. So I think the text is good as is? Explaining the nature of the bug is probably not appropriate for a docs warning. What I would like to do after this is merged is summarize the remaining issues in a single issue, because info is currently spread over quite a few issues and PRs.

Reply:
Maybe it is worth adding that backward is not implemented when `src.shape > index.shape`, which means it only works if `src.shape == index.shape` :(

Reply:
Makes sense. So for `scatter_` we have:

<img width="485" alt="image" src="https://user-images.githubusercontent.com/98330/102908294-877f5c80-4477-11eb-80fc-74ceb0d155ff.png">

And for `scatter_add_` we have:

<img width="485" alt="image" src="https://user-images.githubusercontent.com/98330/102908334-982fd280-4477-11eb-8e7d-6920dfffa9d3.png">

Reply:
@rgommers, I think instead of saying "gradient will be incorrect" you should say that gradient will be propagated to all locations in the source that correspond to the same index. Otherwise it's unclear how incorrect the gradient is - does it contain all `nans`? Does it contain random values?

Reply:
good point, thanks. done